tests/std/tests/GH_002989_nothrow_unwrappable/test.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <filesystem>
+#include <list>
+#include <type_traits>
+#include <vector>
+#include <xutility>
+
+#if _HAS_CXX20
+#include <ranges>
+#endif
+
+using namespace std;
+
+#if _HAS_CXX17
+using filesystem::path;
+#endif
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+struct Predicate {
+    template <class T>
+    bool operator()(const T&) const {
+        return true;
+    }
+};
+
+template <class It, bool CopyUnwrapNothrow = true>
+void do_single_test() {
+    // !a || b is equivalent to a => b (a implies b)
+    // This is written this way to avoid `if constexpr` in C++14 mode.
+    STATIC_ASSERT(_Unwrappable_v<It> == _Is_nothrow_unwrappable_v<It>);
+    STATIC_ASSERT(_Unwrappable_v<It> == _Is_nothrow_unwrappable_v<It&&>);
+    STATIC_ASSERT(noexcept(_Get_unwrapped(declval<It>())));
+
+    STATIC_ASSERT(!_Unwrappable_v<It> || _Is_nothrow_unwrappable_v<const It&> == CopyUnwrapNothrow);
+    STATIC_ASSERT(!_Unwrappable_v<It> || _Is_nothrow_unwrappable_v<const It&&> == CopyUnwrapNothrow);
+    STATIC_ASSERT(noexcept(_Get_unwrapped(declval<const It&>())) == CopyUnwrapNothrow);
+    STATIC_ASSERT(noexcept(_Get_unwrapped(declval<const It&&>())) == CopyUnwrapNothrow);
+}
+
+template <class It, bool CopyUnwrapNothrow = true>
+void do_full_test() {
+    do_single_test<It, CopyUnwrapNothrow>();
+    do_single_test<reverse_iterator<It>, CopyUnwrapNothrow>();
+    do_single_test<move_iterator<It>, CopyUnwrapNothrow>();
+
+#ifdef __cpp_lib_concepts // TRANSITION, GH-395
+    using R = ranges::subrange<It, It>;
+
+    // TRANSITION, GH-2997
+    do_single_test<ranges::iterator_t<ranges::filter_view<R, Predicate>>, true>();
+    // TRANSITION, GH-2997
+    do_single_test<ranges::iterator_t<ranges::transform_view<R, Predicate>>, true>();
+    if constexpr (bidirectional_iterator<It>) {
+        do_single_test<ranges::iterator_t<ranges::reverse_view<R>>, CopyUnwrapNothrow>();
+    }
+#endif // __cpp_lib_concepts
+}
+
+struct BidiIterUnwrapThrowing : vector<int>::iterator {

```

Comment:
No change requested: This is technically non-Standard (`vector<int>::iterator` could be a pointer).

Reply:
We also depend on `vector<int>::iterator` having an `_Unwrapped()` member func 😄